### DISCUSSION TITLE
Update Mechanism to automatically pull updates from GitHub (see #596)

### DISCUSSION THREAD

**Discussion by smileBeda** [bootscore/bootscore#603] [Ideas]
@crftwrk please find here v1-alpha of the updater classes both for theme and plugins:
https://github.com/TukuToi/github-updater/

They should be functional, as I tested during development, but as usual made some minor changes prior to uploading to git, that why currently it is `alpha`.

Description of the intended usage in the readme.

As it is late for me here I will rest and test another round tomorrow also with your more particular setup of `-main` in the names to ensure it all works.
Further I want to:
- [ ] try to use [`update_plugins_{$hostname}`](https://developer.wordpress.org/reference/hooks/update_plugins_hostname/) (for if WP Install > 5.8) in the plugin updater
- [ ] try to narrow down the screens in the theme updater where we check for updates
- [ ] ensure that I truly cover all capability checks as necessary
- [ ] Do some QC in a "test to break" attitude.
- [ ] Do some more due diligence with code style and format - I saw that specially the lsat submitted theme updater had some indents messed up

However, you can already very well see how it would be intended to work, how you would include it in your software, and of course, feedback welcome (there is a lot of additional stuff one can do with GitHub api, like show off related software, download counts, Yada Yada. But I think that is out of scope. Leaner is better).

---

I thought to use this _discussions_ format rather than an issue ticket, so we can also decide how we handle the code, whether you will just clone the repo or if you prefer a PR for each plugin/theme etc etc.

**Comment by crftwrk** [comment]
Created a simple plugin https://github.com/crftwrk/bs-simple-notice and did some releases, but I can't get updates to work. What I'm doing wrong?

**Reply by smileBeda** [reply]
Off the top of my head (sorry mobile right now) your plugin release zip is a -main and your installed one isn’t?
The names must match.

so if you plan to install and update it as -main it has all to be named -main
(Plugin name, main file, folder… all must point to a “-main” plugin name. It appears to me your plugin is only named “main” in the file but “bs-simple-notice” in the folder.
Those differences can trip off wp and the code for updates.

if you already do that I’ll have to check if perhaps org endpoint response is different than single user repo endpoint.

also - but I guess that is not the cause - make sure you either click “check for updates” or go to “wp admin > updates” to clear the transient of wp.

**Reply by smileBeda** [reply]
Hi @crftwrk 
The debug log of the GitHub API request in your plugin use case returns `array(2) { ["message"]=> string(9) "Not Found" ["documentation_url"]=> string(28) "https://docs.github.com/rest" }`

---

There are two problems to fix in the example plugin you made:
- your main plugin file you define the GitHub URL as `https://api.github.com/crftwrk/bs-simple-notice-main`. Instead, it has to be `https://api.github.com/repos/username|org/repo-name` as specified in the https://github.com/TukuToi/github-updater#readme example code. Thus - in your case that becomes `https://api.github.com/repos/crftwrk/bs-simple-notice`
- your release is not tagged "properly". It has to be a valid version (PHP-standardized): `1.0.0` for example. Your tag however is `v1.05`. See also https://www.php.net/manual/en/function.version-compare.php. We could resolve that in the updater by ltrimming the `v` but it is suggested to instead use standardised versioning.
<img width="298" alt="Screenshot 2023-10-16 at 10 23 37" src="https://github.com/bootscore/bootscore/assets/11800236/356783ad-4ed0-4fe6-934c-1340581cb4ef">

---

Once I fixed the GitHub API Url in my local copy of your plugin's main file, and (since I could not create a new tag on your repo of course) ltrimmed the `v` from the tag I received from it in the code, I was able to see and pull your update from GitHub.
<img width="1607" alt="Screenshot 2023-10-16 at 10 21 28" src="https://github.com/bootscore/bootscore/assets/11800236/69d255ae-5872-47d5-8270-15f2bc5fafc3">

---

**Suggestion**
- if you do not already, I suggest a release workflow. Example here https://github.com/TukuToi/better-wp-plugin-boilerplate/blob/develop/.github/workflows/release.yml
- now you might see why I am so adamant about release names and plugin/theme names (regarding those `-main` in the slugs). Having it all unified as per _suggestions_ just saves a lot of confusion.

Note that it is also suggested to have the plugin folder, plugin main file, plugin text domain and plugin name to be all the same. Not mandatory (not in all cases) but strongly suggested by WP itself. There are some edge cases that can cause major havoc when those are not matching up. It is _not_ an issue here. Just underlining the point that it would be "better" to have unified naming everywhere. It gets trickier in themes, where the naming is even more important :)

I do fully understand your reasoning of `-main` since some users use to simply download the code and install. But _strictly speaking_ that is the _users wrongdoing_ and should be corrected via Readme/Support rather than adapting the code to the error. Comparably, when you download _another branch_, that would say `{repo}-{branch-name}`. And in this case you also would not adapt the code to fix the user error, but instead inform the user to download a proper _release_ instead.

**Reply by crftwrk** [reply]
Played around with some libraries and understand your problem with `-main` now. The thing is, that it's impossible for us to remove it, because it will crash the entire system. Unfortunately we have to live with that.

I always end up with [PUC](https://github.com/YahnisElsts/plugin-update-checker). This has the most advanced options such as manage private repos, even it is a larger library and must included in plugin/theme.

[Git Updater](https://github.com/afragen/git-updater) seems to be a good option to manage updates in the backend. The integration in theme/plugin is just a single line and $19,95/year is fair.

**Reply by smileBeda** [reply]
Right. Just to make sure: you do _not_ need to rename it in the current updater proposed at the repo.

I understand this cannot be done anymore at this point as it would require more user awareness which is usually not something a developer can rely on.

To be sure, you will continue using the updater you mentioned?

**Reply by crftwrk** [reply]
Thanks, got it. Installed [v1.4.0](https://github.com/crftwrk/bs-simple-notice/releases/tag/1.4.0) and updated successfully to v1.5.0 with TukuToi updater.

But is there a way to include `-main` suffix in zip file?

Having the `v` in tag is the common usage and recommended by GitHub, but we can skip that. See example  https://github.com/twbs/bootstrap/releases/tag/v5.3.2

> To be sure, you will continue using the updater you mentioned?

No, I'm open to any updater, just testing different options. Think Justin doesn't want to have another larger library in the theme, so simplest way should be the best option.

**Reply by smileBeda** [reply]
Right.

you do not need to remove the -main
As mentioned I got your plugin to update - and it uses -main. As long the suffix _matches_ it’ll work.

the `v-` prefix is not an accepted valid php standard version. Neither a semver, I believe. Not sure why GitHub recommends that. In any case, _if_ you want, it’s easy to update the code I made to consider such prefix (or any prefix). It’s just need an additional function call to “sanitize” the tag we receive from git.

However I’d personally recommend to stick with just x.x.x Format. But as said, no problem at all to make the code a bit more “versatile” towards other formats.
As it was intended for a “tailored” purpose I didn’t add in fallbacks for all possible edge cases, simply to keep it as lean as possible.

**Reply by crftwrk** [reply]
I'm very busy right now and haven't played around with the updater any further. But it seems to work well. However, we first have to check the theme for possible trouble-makers before we implement it. Here are two candidates which have to be solved first:

- https://github.com/bootscore/bootscore/pull/575 has to be merged
- https://github.com/bootscore/bootscore/issues/419 has to be fixed, means links deleted

I think version 5.4 is the earliest version where we can add an auto-updater.
